### PR TITLE
[SYCL][Doc] Add sycl complex extension proposal

### PR DESCRIPTION
**Pull request by AidanBeltonS**
This PR proposes an extension sycl complex. It proposes the introduction of an sycl complex interface which can be adjusted independently to `std::complex`. 

This framework would allow for further development of the complex interface independently of the C++ standard. For example add support for complex `vec` and `marray`.

### CODE REVIEW

**Review by jbrodman — APPROVED** [review]
LGTM

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex.asciidoc [hunk 1]
```diff
@@ -0,0 +1,467 @@
...
+
+    // Note: The below << and >> operator with basic_ostream and basic_istream
+    // are host-only.
+    template<class T, class charT, class traits>
+    std::basic_ostream<charT, traits>&
+    operator<<(std::basic_ostream<charT, traits>&, const complex<T>&);
+
+    template<class T, class charT, class traits>
+    std::basic_istream<charT, traits>&
+    operator>>(std::basic_istream<charT, traits>&, const complex<T>&);
```
Comment (steffenlarsen) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex.asciidoc:223]: To my knowledge we don't define these for other SYCL types, so I don't think we should for this either. It also creates a hard dependency on `istream` and `ostream` which we are currently trying to reduce.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex.asciidoc [hunk 2]
```diff
@@ -0,0 +1,467 @@
...
+    template<> class complex<double>;
+    template<> class complex<float>;
+    template<> class complex<sycl::half>;
+
+} // namespace oneapi
+} // namespace ext
+```
+
+The `sycl::oneapi::ext::complex` specializations can be generalised similar
+to existing sycl arithmetic types. The generic type `gencomplex` is defined as
```
Comment (steffenlarsen) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex.asciidoc:245]:
```suggestion
to existing SYCL arithmetic types. The generic type `gencomplex` is defined as
```

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex.asciidoc [hunk 3]
```diff
@@ -0,0 +1,467 @@
...
+types `complex<double>`, `complex<float>`, `complex<sycl::half>`.
+
+The table belows shows the operators defined by the SYCL complex interface
+along with a description of its operation.
+
+Note: When performing operations between complex numbers and decimal's.
+The decimal is treated as a complex number with a real component equal to
```
Comment (steffenlarsen) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex.asciidoc:252]:
```suggestion
Note: When performing operations between complex numbers and decimals,
the decimal is treated as a complex number with a real component equal to
```

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex.asciidoc [hunk 4]
```diff
@@ -0,0 +1,467 @@
...
+|`std::basic_ostream& operator<<(std::basic_ostream& x, const gencomplex& y);`
+|Streams the complex number y in the format "(real,imaginary)" into `std::basic_ostream` x and return the result. This is only available for the host.
+|`std::basic_istream& operator>>(std::basic_istream&, const gencomplex& y);`
+|Streams the string given in the format "(real,imaginary)" into complex number y. This is only available for the host.
+|===
+
+
+=== Mathematical operations
+
+This proposal adds `ext::oneapi` namespace math functions accepting
```
Comment (steffenlarsen) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex.asciidoc:324]:
```suggestion
This proposal adds `sycl::ext::oneapi` namespace math functions accepting
```

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex.asciidoc [hunk 5]
```diff
@@ -0,0 +1,467 @@
...
+|Compute the tangent of complex number x.
+|`gencomplex tanh(const gencomplex& x)`
+|Compute the hyperbolic tangent of complex number x.
+|===
+
+== Implementation notes
+
+The complex mathematical operations can all be defined using SYCL built-ins.
+Therefore, implementing complex with SYCL built-ins would allow any backend
+with SYCL built-ins to support `ext::oneapi::complex`. The current
```
Comment (steffenlarsen) [inline sycl/doc/extensions/proposed/sycl_ext_oneapi_complex.asciidoc:453]:
```suggestion
with SYCL built-ins to support `sycl::ext::oneapi::complex`. The current
```

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex.asciidoc [hunk 6]
```diff
@@ -0,0 +1,467 @@
...
+            complex(const complex&)
+            complex(const std::complex<T>&)
+
+            operator std::complex();
+
+            T real() const {return re;}
+            T imag() const {return im;}
+
+            void real(T _re) {re = _re;}
+            void imag(T _im) {im = _im;}
```
[the remark this answers is not in the document]
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex.asciidoc]: Thank you for pointing this out. I have removed the implementation.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex.asciidoc [hunk 7]
```diff
@@ -0,0 +1,467 @@
...
+            // operator-=
+            template<X>
+            complex& operator-=(const complex<X>&);
+
+            // operator*=
+            template<X>
+            complex& operator*=(const complex<X>&);
+
+            // operator/=
+            template<X>
```
[the remark this answers is not in the document]
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex.asciidoc]: I have added the missing `typename` to the operator assigns.

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex.asciidoc [hunk 8]
```diff
@@ -0,0 +1,467 @@
...
+The core of this extension in the complex math class. This class contains a real
+and imaginary component and enables mathematical operations between complex
+numbers and decimals. The complex class interface and operations are shown
+below.
+
+The complex type is trivially copyable and type trait `is_device_copyable`
+should resolve to `std::true_type`.
+
+```C++
+namespace ext {
```
[the remark this answers is not in the document]
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex.asciidoc:100]: Okay, I have added the `sycl` namespace to this and other interface definitions

File: sycl/doc/extensions/proposed/sycl_ext_oneapi_complex.asciidoc (hunk 1 above)
Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex.asciidoc:224]: Thank you for the clarification. I have removed the `istream` and `ostream` operators, so there is no hard dependency in the extension proposal.

Comment (steffenlarsen) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex.asciidoc:224]: They seem to still be here. Did you forget to push them or am I missing something?

Comment (AidanBeltonS) [reply sycl/doc/extensions/proposed/sycl_ext_oneapi_complex.asciidoc]: Sorry I forgot to push the change.

**Review by steffenlarsen — APPROVED** [review]
LGTM!

**Pull request merged**